### PR TITLE
Add 8051 disassemble/reassemble tests on realtek firmware.

### PR DESCRIPTION
**Pull request by gogo2464**
- [X] Mark this if you consider it ready to merge
- [X] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

### CODE REVIEW

File: test/Makefile [hunk 1]
```diff
@@ -50,6 +62,8 @@ vc4: bins
 yara: bins
 	${RUNTEST} db/extras/cmd/yara
 
+
```
Comment (trufae) [inline test/Makefile:65]: delete those empty lines

File: test/Makefile [hunk 2]
```diff
@@ -17,6 +17,18 @@ RUNTEST=r2r -L -o results.json
 r2r-tests: bins
 	${RUNTEST}
 
+8051: bins
+	radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > MP_8192E_FW_NIC.asm
+	rasm2 -a 8051 -o 0x8000 -f MP_8192E_FW_NIC.asm -B  > recompiled-firmware.bin
+	if [ "$(diff -s ./bins/8051/MP_8192E_FW_NIC.bin recompiled-firmware.bin)" != "Binary files ./bins/8051/MP_8192E_FW_NIC.bin and recompiled-firmware.bin differ" ]
```
Comment (trufae) [inline test/Makefile:23]: use `cmp` instead. `diff` is not portable

File: test/Makefile [hunk 3]
```diff
@@ -17,6 +17,18 @@ RUNTEST=r2r -L -o results.json
 r2r-tests: bins
 	${RUNTEST}
 
+8051: bins
+	radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > MP_8192E_FW_NIC.asm
```
Comment (trufae) [inline test/Makefile:21]: make clean should remove this .asm file

File: test/Makefile [hunk 4]
```diff
@@ -17,6 +17,18 @@ RUNTEST=r2r -L -o results.json
 r2r-tests: bins
 	${RUNTEST}
 
+8051: bins
```
Comment (trufae) [inline test/Makefile:20]: run this target from another make target or in the ci, so we can see if the test works before mergin

**Review by trufae — CHANGES_REQUESTED** [review]
.

File: test/Makefile (hunk 4 above)
Comment (gogo2464) [reply test/Makefile:20]: I am 100% sure it works. When I test locally it works. I also pushed a very similar test on another repo. See: https://github.com/gogo2464/realtek-firmware-decompilation/blob/master/.github/workflows/.travis.yml#L24

Comment (gogo2464) [reply test/Makefile:20]: Ah yes I just mentionned the issue with another architecture.

**Comment by gogo2464** [comment]
The http file takes too long I am not responsible of that.

File: Makefile [hunk 5]
```diff
@@ -391,6 +391,7 @@ shot:
 		radare.org:/srv/http/radareorg/get/shot
 
 tests test:
+	export LC_ALL=C
```
Comment (trufae) [inline Makefile:394]: That line those nothing. The very next line wont maintain the same environment

also its wrong to override user settings. If you have warnjngs fix your system

File: test/Makefile (hunk 4 above)
Comment (trufae) [reply test/Makefile:20]: Well. You cant be sure if that will break tomorrow. Theres no sense in adding tests that never run

**Review by trufae — CHANGES_REQUESTED** [review]
Which http are you talking about

File: Makefile (hunk 5 above)
Comment (gogo2464) [reply Makefile:394]: It is to ensure that local tests of everyone running tests have same setting.

**Comment by gogo2464** [comment]
By http I mean the failling test.

File: Makefile (hunk 5 above)
Comment (trufae) [reply Makefile:394]: no. this line is useless. try this if my epxlanation above wasnt clear

```
$ make
export JEJE=123
echo $JEJE

$ cat Makefile
all:
	export JEJE=123
	echo $$JEJE
$
```

File: test/Makefile [hunk 6]
```diff
@@ -17,6 +17,18 @@ RUNTEST=r2r -L -o results.json
...
+	radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > ./tmp/MP_8192E_FW_NIC.asm
+	rasm2 -a 8051 -o 0x8000 -f ./tmp/MP_8192E_FW_NIC.asm -B  > ./tmp/recompiled-firmware.bin
+	if [ "$(cmp ./bins/8051/MP_8192E_FW_NIC.bin ./tmp/recompiled-firmware.bin)" != "./bins/8051/MP_8192E_FW_NIC.bin ./tmp/recompiled-firmware.bin differ:"* ]
```
Comment (trufae) [inline test/Makefile:23]: thats not valid makefile. please read the other comments and address them

**Comment by trufae** [comment]
The simplest way to move on here is just to use a shellscript in `test/test-8051-reassembly.sh` (and call it from the makefile). because makefiles are not meant to be used as shellscript replacements. the logic is different

File: Makefile (hunk 5 above)
Comment (gogo2464) [reply Makefile:394]: ah ok

File: test/Makefile (hunk 3 above)
Comment (gogo2464) [reply test/Makefile]: Nice catch. I decided to put the .asm file in the ./tmp folder to delete it.

**Comment by gogo2464** [comment]
Hard to test with no master running.

**Comment by trufae** [comment]
the ci is green again you can rebase if you like the green color

**Comment by gogo2464** [comment]
I prefer the red color. But I also prefer when the code is running. Then let's rebase. :)

File: test/test-8051-reassembly.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+mkdir -p ../tmp
```
Comment (trufae) [inline test/test-8051-reassembly.sh:1]:
```
#!/bin/sh

T=$(mktemp -d tmp.XXXXXX)
[ -z "$T" ] && exit 1
mkdir -p $T
....

rm -rf $T
```

File: test/test-8051-reassembly.sh [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > ../tmp/MP_8192E_FW_NIC.asm
+rasm2 -a 8051 -o 0x8000 -f ../tmp/MP_8192E_FW_NIC.asm -B  > ../tmp/recompiled-firmware.bin
+if [ "$(cmp ./bins/8051/MP_8192E_FW_NIC.bin ../tmp/recompiled-firmware.bin)" != "./bins/8051/MP_8192E_FW_NIC.bin ../tmp/recompiled-firmware.bin differ:"* ]
```
Comment (trufae) [inline test/test-8051-reassembly.sh:5]: just compare the return code. the output may differ between implementations

```
if [ $? = 0 ]; then ,...

```

File: test/test-8051-reassembly.sh [hunk 9]
```diff
@@ -0,0 +1,12 @@
+mkdir -p ../tmp
+
+radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > ../tmp/MP_8192E_FW_NIC.asm
```
Comment (trufae) [inline test/test-8051-reassembly.sh:3]: check return code with `|| exit 1`

File: test/test-8051-reassembly.sh [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+rasm2 -a 8051 -o 0x8000 -f ../tmp/MP_8192E_FW_NIC.asm -B  > ../tmp/recompiled-firmware.bin
```
Comment (trufae) [inline test/test-8051-reassembly.sh:4]: check error code with || exit 1

File: test/test-8051-reassembly.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
+mkdir -p ../tmp
+
```
Comment (trufae) [inline test/test-8051-reassembly.sh:2]:
```suggestion
printf "Reassembling 8051... "

```

File: test/test-8051-reassembly.sh [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > ../tmp/MP_8192E_FW_NIC.asm
+rasm2 -a 8051 -o 0x8000 -f ../tmp/MP_8192E_FW_NIC.asm -B  > ../tmp/recompiled-firmware.bin
+if [ "$(cmp ./bins/8051/MP_8192E_FW_NIC.bin ../tmp/recompiled-firmware.bin)" != "./bins/8051/MP_8192E_FW_NIC.bin ../tmp/recompiled-firmware.bin differ:"* ]
+then
+	echo "OK"
+else
+	radiff2 -q -D ./bins/8051/MP_8192E_FW_NIC.bin ../tmp/recompiled-firmware.bin
+	echo "failed"
+	exit -1
```
Comment (trufae) [inline test/test-8051-reassembly.sh:11]:
```suggestion
	exit 1
```

File: test/test-8051-reassembly.sh [hunk 13]
```diff
@@ -0,0 +1,12 @@
+mkdir -p ../tmp
+
+radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > ../tmp/MP_8192E_FW_NIC.asm
+rasm2 -a 8051 -o 0x8000 -f ../tmp/MP_8192E_FW_NIC.asm -B  > ../tmp/recompiled-firmware.bin
+if [ "$(cmp ./bins/8051/MP_8192E_FW_NIC.bin ../tmp/recompiled-firmware.bin)" != "./bins/8051/MP_8192E_FW_NIC.bin ../tmp/recompiled-firmware.bin differ:"* ]
+then
+	echo "OK"
```
Comment (trufae) [inline test/test-8051-reassembly.sh:7]:
```suggestion
	echo "OK"
	exit 0
```

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by gogo2464** [comment]
Ready to merge but it would be better if I did not disassemble/reassemble the offset.

File: test/Makefile [hunk 14]
```diff
@@ -1,4 +1,8 @@
-all: r2r-tests unit_tests
+all:
+	export LC_ALL=C
```
Comment (trufae) [inline test/Makefile:2]:
```suggestion
```

File: test/Makefile [hunk 15]
```diff
@@ -1,4 +1,8 @@
...
+	$(MAKE) r2r-tests
+	$(MAKE) unit_tests
+	$(MAKE) test-8051-reassembly
```
Comment (trufae) [inline test/Makefile:5]: why dont run those 3 targets in parallel as it was doing before? please change this

File: test/Makefile [hunk 16]
```diff
@@ -17,6 +21,9 @@ RUNTEST=r2r -L -o results.json
 r2r-tests: bins
 	${RUNTEST}
 
+test-8051-reassembly:
+	sh ./test-8051-reassembly.sh
```
Comment (trufae) [inline test/Makefile:25]:
```suggestion
	$(SHELL) ./test-8051-reassembly.sh
```

File: test/test-8051-reassembly.sh [hunk 17]
```diff
@@ -0,0 +1,20 @@
+BUILD_DIRECTORY=$(mktemp -d tmp.XXXXXXXXXX)
+[ -z "$BUILD_DIRECTORY" ] && exit 1
+mkdir -p $BUILD_DIRECTORY
```
Comment (trufae) [inline test/test-8051-reassembly.sh:3]: mktemp already creates the directory, this line is not necessary
```suggestion
```

File: test/test-8051-reassembly.sh [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+printf "Reassembling 8051... "
+radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > $BUILD_DIRECTORY/MP_8192E_FW_NIC.asm || exit 1
+rasm2 -a 8051 -o 0x8000 -f $BUILD_DIRECTORY/MP_8192E_FW_NIC.asm -B  > $BUILD_DIRECTORY/recompiled-firmware.bin || exit 1
+
+cmp ./bins/8051/MP_8192E_FW_NIC.bin $BUILD_DIRECTORY/recompiled-firmware.bin
+if [ $? -eq 0 ]
```
Comment (trufae) [inline test/test-8051-reassembly.sh:10]:
```suggestion
if [ $? -eq 0 ]; then
```

File: test/test-8051-reassembly.sh [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+then
```
Comment (trufae) [inline test/test-8051-reassembly.sh:11]:
```suggestion
```

File: test/test-8051-reassembly.sh [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+	echo "OK"
+	exit 0
+else
+	radiff2 -q -D ./bins/8051/MP_8192E_FW_NIC.bin "$BUILD_DIRECTORY/recompiled-firmware.bin"
+	echo "failed"
+	exit 1
+fi
+
+rm -rf $BUILD_DIRECTORY
```
Comment (trufae) [inline test/test-8051-reassembly.sh:20]: This line is never executed because you exit before leaving. use RC=0 and then exit $RC

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by gogo2464** [comment]
A bit different than expected. You could merge anyway.

**Comment by trufae** [comment]
Stop saying me to merge because your script is not even able to execute

File: test/test-8051-reassembly.sh [hunk 21]
```diff
@@ -0,0 +1,21 @@
+BUILD_DIRECTORY=$(mktemp -d tmp.XXXXXXXXXX)
+[ -z "$BUILD_DIRECTORY" ] && exit 1
+export EXIT_CODE = 0
```
Comment (trufae) [inline test/test-8051-reassembly.sh:3]:
```suggestion
export EXIT_CODE=0
```

**Comment by trufae** [comment]
I think it should be fine now. Lets wait for the ci. But if the script fails to run because the reassembling is not correct then i dont see the point on having this test

File: test/test-8051-reassembly.sh [hunk 22]
```diff
@@ -0,0 +1,21 @@
+BUILD_DIRECTORY=$(mktemp -d tmp.XXXXXXXXXX)
+[ -z "$BUILD_DIRECTORY" ] && exit 1
+export EXIT_CODE=0
+
+printf "Reassembling 8051... "
+radare2 -a 8051 -m 0x8000 -e "scr.color=0" -qf -c "pI" ./bins/8051/MP_8192E_FW_NIC.bin > $BUILD_DIRECTORY/MP_8192E_FW_NIC.asm || exit 1
+rasm2 -a 8051 -o 0x8000 -f $BUILD_DIRECTORY/MP_8192E_FW_NIC.asm -B  > $BUILD_DIRECTORY/recompiled-firmware.bin || exit 1
+
+cmp ./bins/8051/MP_8192E_FW_NIC.bin $BUILD_DIRECTORY/recompiled-firmware.bin
+if [ $? -eq 0 ]
```
Comment (trufae) [inline test/test-8051-reassembly.sh:10]:
```suggestion
if [ $? -eq 0 ]; then
```

**Review by trufae — COMMENTED** [review]
will be good if you try to run the script before pushing a change. or at least apply my suggestions instead of adding more breaking changes

**Pull request merged**